### PR TITLE
Select and insert panel

### PR DESCRIPTION
**Pull request by PatrickMassot**
This PR brings infrastructure to easily define tactics allowing users to select things in the info
view and generate a tactic call. The infrastructure is illustrated by a simpler version of the conv demo. Note that most of the new demo is code from the old one. Mathlib uses will include a `gcongr`  widget and a `congrm` widget when this tactic will be merged. 

Warning: this adds a dependency on `Std`. See the discussion [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Multi.20commands.20macro/near/368162315).

### CODE REVIEW

File: ProofWidgets/Component/SelectInsertPanel.lean [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        let lctx := md.lctx |>.sanitizeNames.run' {options := (← getOptions)}
+        Meta.withLCtx lctx md.localInstances do
+          let insertedCode ← mkCmdStr selectedLocations md.type
+          let textDocumentEdit : Lsp.TextDocumentEdit := {
+            textDocument := { uri := doc.meta.uri, version? := doc.meta.version },
+            edits        := #[{range := replaceRange, newText := insertedCode }] }
+          return some { content := insertedCode,
+                        edit := Lsp.WorkspaceEdit.ofTextDocumentEdit textDocumentEdit,
+                        /- CHECKME: the new cursor position computation below looks very naive.
+                           Will it work with non-ascii insertedCode? -/
```
Comment (PatrickMassot) [inline ProofWidgets/Component/SelectInsertPanel.lean:68]: Let me highlight there is a question here. @Vtec234

**Comment by jcommelin** [comment]
@Vtec234 What's the status here? I would love to play with this in mathlib, and in demo talks that I give.

File: ProofWidgets/Component/SelectInsertPanel.lean (hunk 1 above)
Comment (Vtec234) [reply ProofWidgets/Component/SelectInsertPanel.lean]: Fixed during refactor, though note the cursor may still end up mispositioned due to leanprover/vscode-lean4#225.

**Comment by Vtec234** [comment]
Overall, it seems to me that what you are going for here is not necessarily a general method of defining 'select&insert panels', but rather a way to program some part of a widget in Lean. So instead, I am proposing an alternative API which does this for arbitrary widgets: we write a function `Props -> RequestM (RequestTask Html)` and get out a `Component Props`. With this your entire demo can be written in Lean: please see `SelectInsertConv2.lean` for an example usage.

If you think this is a good API for your uses, I would suggest either dropping `SelectInsertPanel.lean` entirely, or shortening it to just one function that deals with the boilerplate, probably of type `(title : String) (help : String) (mkCmdStr : Array Lean.SubExpr.GoalsLocation -> Expr -> MetaM String) : SelectionPanelProps -> RequestM (RequestTask Html)`, and then using the new API.

File: ProofWidgets/Component/Panel/SelectInsertPanel.lean [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+the given callback function to generate a String that will replace the tactic when
+users click on a link.  The callback function must have type
+`(pos : Array Lean.SubExpr.GoalsLocation) → (goalType : Expr) → MetaM String`. -/
+macro "mkSelectInsertTactic" name:str title:str help:str callback:ident : command =>
+  let cmdName : Name := .str `mkSelectInsertCmd (name.getString ++ "_widget")
+  `(
+    @[server_rpc_method]
+    def $(mkIdent cmdName) := mkSelectInsertCommand $callback
+    elab stx:$name:str : $(mkIdent `tactic) =>
+      mkInsertTactic ``mkInsertWidget $title:str $help:str $(quote cmdName) stx)
```
Comment (Vtec234) [inline ProofWidgets/Component/Panel/SelectInsertPanel.lean:113]: I am not a fan of macro-generating macros like this. I am wary of ending up in a situation where our code cannot be read without reading a bunch of macro implementations first, thus adding a prerequisite of 'user knows metaprogramming'. I would prefer to stay closer to object-level whenever possible, even if that means writing a bit more boilerplate every time: I mean something like for example `SelectInsertConv2.lean` where I am tagging the RPC method by hand.

Comment (Vtec234) [reply ProofWidgets/Component/Panel/SelectInsertPanel.lean:113]: On the other hand one could argue that just a `mkSelectInsertTactic` is more readable if one is willing to treat it as a black box! But in that case we should make it a bit more general so as to not have a bunch of `mkDoFooTactic` macros for every `Foo` that one might want to do.

**Pull request merged**